### PR TITLE
Extract discovery package

### PR DESCRIPTION
**Pull request by fabxc**
@brian-brazil 

This moves a few things around. Mainly the TargetSet, which collects updates from different TargetProviders is now public.
This required a bit of untangling to separate it from the scrapePools. Both are now brought together by the TargetManager. This will allow us to hook in different logic to get Alertmanager targets.

I think this also made a few things a bit simpler to understand.

### CODE REVIEW

**Comment by tomwilkie** [comment]
I was just working on something similar: https://github.com/tomwilkie/prometheus/tree/decouple-scraping-and-storage

I wanted to reuse the discover, target manager etc in another project, but provide my own callback for scraping.

**Comment by fabxc** [comment]
Yea, that sounds useful too and from a first glance, the change here probably won't impact the changes you made at all.
For alerting the definition of a scrape target doesn't fully match the idea of an Alertmanager target. So this hooks in one layer below.

File: discovery/gce/gce.go [hunk 1]
```diff
@@ -52,15 +52,13 @@ const (
 var (
 	gceSDRefreshFailuresCount = prometheus.NewCounter(
 		prometheus.CounterOpts{
-			Namespace: namespace,
```
Comment (brancz) [inline discovery/gce/gce.go]: 👍 for the change. I think you removed all usages of this in this PR so we can remove the [var](https://github.com/prometheus/prometheus/blob/d67dd561af2cd01d08724c9007d86603891d1261/discovery/gce/gce.go#L49) all together.

**Comment by brancz** [comment]
Rather large change so I haven't looked at everything in detail yet, but overall looks good. I'm gonna give it a spin and report back. Besides that I really like this change as it makes a very nice reusable component/package.

File: retrieval/targetmanager.go [hunk 2]
```diff
@@ -111,23 +98,33 @@ func (tm *TargetManager) reload() {
...
+				ts.ts.Run(ctx)
+				ts.sp.stop()
 				tm.wg.Done()
 			}(ts)
 		} else {
-			ts.reload(scfg)
+			ts.sp.reload(scfg)
 		}
-		ts.runProviders(tm.ctx, providersFromConfig(scfg))
+		ts.ts.UpdateProviders(discovery.ProvidersFromConfig(scfg))
```
Comment (brancz) [inline retrieval/targetmanager.go]: Haven't quite figured out why, but this panics 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b64e2]

goroutine 164 [running]:
panic(0x1891220, 0xc420016020)
        /usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:500 +0x1a1
github.com/prometheus/prometheus/retrieval.(*TargetManager).reload(0xc42044b880)
        /Users/fredericbranczyk/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:123 +0x132
github.com/prometheus/prometheus/retrieval.(*TargetManager).Run(0xc42044b880)
        /Users/fredericbranczyk/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:68 +0x13e
created by main.Main
        /Users/fredericbranczyk/go/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:213 +0x165f
```

Comment (fabxc) [inline retrieval/targetmanager.go]: Accidentally used `:=` instead of `=` in the if condition. Fix pushed.

**Comment by brancz** [comment]
Works for me. Will now do a more in depth review...

File: retrieval/scrape.go [hunk 3]
```diff
@@ -206,11 +201,32 @@ func (sp *scrapePool) reload(cfg *config.ScrapeConfig) {
 	)
 }
 
+// Sync converts target groups into actual scrape targets and snychronizes
```
Comment (brancz) [inline retrieval/scrape.go]: typo "snychronizes"

**Comment by brancz** [comment]
lgtm 👍

**Comment by brancz** [comment]
But given the size of the PR I think another pair of eyes would be good

File: discovery/discovery.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+		case <-ts.syncCh:
+			ts.sync()
+		case p := <-ts.providerCh:
+			ts.updateProviders(ctx, p)
+		}
+	}
+}
+
+func (ts *TargetSet) sync() {
+	ts.mtx.RLock()
```
Comment (brian-brazil) [inline discovery/discovery.go]: Should we drain ts.syncCh here in case several have backed up? Having to wait 5s for every pending update would be ungood.

**Comment by fabxc** [comment]
Got +1 out-of-band. Merging.

**Pull request merged**